feat(Microsoft Teams Node): Add support for government cloud tenants

antman1p opened the pull request:
## Summary

This PR adds configurable Microsoft Graph API base URL support to the Microsoft Teams node, allowing users to connect to government cloud tenants (GCC/DoD) that use different Graph API endpoints.

**Changes made:**
- Added `baseUrl` field to `MicrosoftOAuth2Api` credentials with default value `https://graph.microsoft.com`
- Modified `microsoftApiRequest` function in Teams v2 transport to use the configurable base URL
- Maintains backward compatibility with existing public cloud tenants

**Testing:**
1. Create Microsoft Teams credentials with custom base URL (e.g., `https://graph.microsoft.us` for GCC)
2. Use Microsoft Teams node to list channels or send messages
3. Verify successful connection to government cloud tenants

**Before:** Hardcoded `https://graph.microsoft.com` caused "401 - InvalidCloudInstance" errors for GCC/DoD tenants
**After:** Configurable base URL supports all Microsoft Graph instances including government clouds

## Related Linear tickets, Github issues, and Community forum posts

Fixes [#17283](https://github.com/n8n-io/n8n/issues/17283) - Microsoft Teams node fails with "401 - InvalidCloudInstance" for GCC/DoD tenants

## Review / Merge checklist

- [x] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) <!--
   **Remember, the title automatically goes into the changelog.
   Use `(no-changelog)` otherwise.**
-->
- [x] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created. [PR #4126 Documentation: Microsoft Graph API Government Cloud Support](https://github.com/n8n-io/n8n-docs/pull/4126)
- [x] Tests included. <!--
   A bug is not considered fixed, unless a test is added to prevent it from happening again.
   A feature is not complete without tests.
-->
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Adds a configurable Microsoft Graph API base URL in credentials and updates Microsoft nodes to use it instead of hardcoded endpoints, with comprehensive tests.
> 
> - **Credentials**
>   - Add `graphApiBaseUrl` option to `MicrosoftOAuth2Api` with presets for Global, US Gov, US Gov DOD, and China.
> - **Microsoft nodes: use configurable base URL (strip trailing slashes, default to global)**
>   - `Microsoft/Teams v2`: `microsoftApiRequest` now builds URLs from `graphApiBaseUrl`.
>   - `Microsoft/Outlook v2`: `microsoftApiRequest` builds URLs from `graphApiBaseUrl`, including shared mailbox paths.
>   - `Microsoft/OneDrive`: `microsoftApiRequest` and `getPath` use `graphApiBaseUrl`.
>   - `Microsoft/Excel v2`: `microsoftApiRequest` uses `graphApiBaseUrl`.
>   - `Microsoft/Entra`: `microsoftApiRequest` and `microsoftApiPaginateRequest` use `graphApiBaseUrl`.
>   - `Microsoft/GraphSecurity`: `msGraphSecurityApiRequest` uses `graphApiBaseUrl`.
> - **Tests**
>   - Add unit tests for each node to verify base URL selection, default fallback, and trailing-slash handling (including GCC/DoD/China endpoints).
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 2b8abb6ea5bafe92bc8705e6d1216a518abcf8eb. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

antman1p committed:
Added GraphAPI endpoint url to credentials

antman1p committed:
Fixed build error on relace() function

antman1p committed:
Update packages/nodes-base/nodes/Microsoft/Teams/v2/transport/index.ts

Fixed assertion per suggestion

Co-authored-by: cubic-dev-ai[bot] <191113872+cubic-dev-ai[bot]@users.noreply.github.com>

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

Joffcom commented on packages/nodes-base/credentials/MicrosoftOAuth2Api.credentials.ts:
```diff
@@ -33,6 +33,16 @@ export class MicrosoftOAuth2Api implements ICredentialType {
+		{
+			displayName: 'Microsoft Graph API Base URL',
+			name: 'baseUrl',
+			type: 'string',
```
This should probably be a dropdown / options field with the options being available so that users can't input the incorrect value.

Does this change also impact oauth or is that always the same url no matter what instance is used?

antman1p replied:
That hardcoded URL is always the same and even when you connect to a non-public registered Azure app it breaks when you try to send an API call, because it always sends to the public tenant URL.

antman1p replied:
I changed it from user input to a dropdown selection for the 4 published graph api endpoints available.  One public and 3 national.  https://learn.microsoft.com/en-us/graph/deployments

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

Joffcom commented:
Hey @antman1p,

Pretty much the same as the other one although I am still not keen on that text field approach.

Joffcom commented:
Ah no I lied... This one doesn't need Dana it is with @RomanDavydchuk to review so it will happen as soon as he can fit it in.

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Changed graph API endpoint to dropdown selection

antman1p commented:
> Hey @antman1p,
> 
> Pretty much the same as the other one although I am still not keen on that text field approach.

I changed it from user input to a dropdown selection for the 4 published graph api endpoints available.  One public and 3 national.  https://learn.microsoft.com/en-us/graph/deployments

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p commented:
Hi @Joffcom, I just wanted to follow up on this to see if it has been assigned.

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

RomanDavydchuk reviewed (changes requested):
Hi, thanks for your contribution!

I looked through your PR and left some comments in the code. Also, while I did test it manually with the global endpoint (`https://graph.microsoft.com`), I don't have access to tenants that use other endpoints, so I'm going to trust your comments where you told us that it's working.

Also, we require PRs to include tests (where it makes sense), so could you please add tests for the `microsoftApiRequest` function?

antman1p committed:
Update packages/nodes-base/nodes/Microsoft/Teams/v2/transport/index.ts

Changed unnecessary cast

Co-authored-by: RomanDavydchuk <roman.d@radency.com>

antman1p commented on packages/nodes-base/credentials/MicrosoftOAuth2Api.credentials.ts:
```diff
@@ -33,6 +33,33 @@ export class MicrosoftOAuth2Api implements ICredentialType {
 			type: 'string',
 			default: 'https://login.microsoftonline.com/common/oauth2/v2.0/token',
 		},
+		{
+			displayName: 'Microsoft Graph API Base URL',
+			name: 'baseUrl',
+			type: 'options',
+			options: [
+				{
+					name: 'Global (https://graph.microsoft.com)',
+					value: 'https://graph.microsoft.com',
+				},
+				{
+					name: 'US Government (https://graph.microsoft.us)',
+					value: 'https://graph.microsoft.us',
+				},
+				{
+					name: 'US Government DOD (https://dod-graph.microsoft.us)',
+					value: 'https://dod-graph.microsoft.us',
+				},
+				{
+					name: 'China (https://microsoftgraph.chinacloudapi.cn)',
+					value: 'https://microsoftgraph.chinacloudapi.cn',
+				},
+			],
+			default: 'https://graph.microsoft.com',
+			required: true,
+			description:
+				'The base URL for the Microsoft Graph API. Select the correct endpoint for your cloud environment.',
+		},
```
Entra, OneDrive, Outlook, Teams, and Excel all use the graphapi.  https://github.com/search?q=org%3An8n-io%20https%3A%2F%2Fgraph.microsoft.com&type=code

RomanDavydchuk replied:
While that is true, all those nodes need to be updated, since they have `https://graph.microsoft.com` hard-coded as the base URL and won't use this dropdown automatically. So leaving this for all Microsoft credentials would be confusing for users, since it'll work only with MS Teams. I see two possible solutions here:
1. Support other base URLs in all other nodes
2. Add the ability to select the base URL only to the MS Teams credentials

antman1p replied:
I made the changes.  I will do functional tests with the government graph API endpoint tomorrow.

antman1p replied:
@Joffcom v23.5.0

antman1p replied:
The reset worked!  Thank you!

antman1p replied:
@RomanDavydchuk @Joffcom I have completed functional test on these changes on the government graph Api tenant, and it works as expected.  I will make the changes to the other Microsoft nodes that require the graph Api this evening.

antman1p replied:
@RomanDavydchuk changes complete.  Ready for your review.

antman1p commented:
> Hi, thanks for your contribution!
> 
> I looked through your PR and left some comments in the code. Also, while I did test it manually with the global endpoint (`https://graph.microsoft.com`), I don't have access to tenants that use other endpoints, so I'm going to trust your comments where you told us that it's working.
> 
> Also, we require PRs to include tests (where it makes sense), so could you please add tests for the `microsoftApiRequest` function?

Thank you.  Yes, it works perfectly.  I have been using it for weeks now without issue.  I will attempt to add the tests.

antman1p committed:
Added tests

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

RomanDavydchuk commented:
> I added tests. Please let me know if these are what you are looking for.

Thanks, but what I actually meant is adding tests for the `microsoftApiRequest` function itself, so you'd test whether it actually uses the base URL from the credentials, falls back to the default one, it also has regex so it might be worth testing that, etc

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Removed tests

antman1p committed:
Removed Credential Graph API dropdown and added it to the Teams node

RomanDavydchuk reviewed (changes requested):
Hey again, looks like you began to update other Microsoft nodes to also allow for configurable base URL, however this is not what I meant with my previous review comment.

Two possible suggestions I had were:
- Keep the "Microsoft Graph API Base URL" property in base Microsoft OAuth2 credentials and update all the nodes that use them to make sure they take this option into account instead of using a hardcoded base URL
- Move the "Microsoft Graph API Base URL" property to Microsoft Teams OAuth2 credentials only, since that was the only node that used it at that point in time

It looks like in your new approach you're adding `graphApiBaseUrl` to every Microsoft _node_. There are a few reasons why it's not the best approach:
- In general, if a URL for a particular service can be changed, that is configured through the credentials (for example, see [Jira Cloud credentials](https://github.com/antman1p/n8n/blob/c4492a65576abd50bd9c937bbc0908b15e1a8536/packages/nodes-base/credentials/JiraSoftwareCloudApi.credentials.ts#L32))
- Making it a property in base credentials is good because it avoid code duplication. If you add it to the node itself, there is a lot of the same code
- Making it a property in credentials is also better because the user doesn't need to specify the base URL in every node, which simplifies UX

I'd suggest you to pick one of the two approaches I mentioned above - unfortunately I won't be able to merge the PR in this current state

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Reverted back to using OAuth Credential graph url dropdown slection

antman1p committed:
chore: Remove accidentally committed git log artifact

antman1p committed:
Fixed Missing Base URL credential usage code and added tests

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p commented:
@RomanDavydchuk are you able to review please?

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Update packages/nodes-base/nodes/Microsoft/Outlook/v2/transport/index.ts

Co-authored-by: cubic-dev-ai[bot] <191113872+cubic-dev-ai[bot]@users.noreply.github.com>

antman1p committed:
Update packages/nodes-base/nodes/Microsoft/Excel/v2/transport/index.ts

Co-authored-by: cubic-dev-ai[bot] <191113872+cubic-dev-ai[bot]@users.noreply.github.com>

antman1p committed:
Update packages/nodes-base/nodes/Microsoft/ToDo/GenericFunctions.ts

Co-authored-by: cubic-dev-ai[bot] <191113872+cubic-dev-ai[bot]@users.noreply.github.com>

antman1p committed:
Update packages/nodes-base/nodes/Microsoft/GraphSecurity/GenericFunctions.ts

Co-authored-by: cubic-dev-ai[bot] <191113872+cubic-dev-ai[bot]@users.noreply.github.com>

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Update packages/nodes-base/nodes/Microsoft/OneDrive/GenericFunctions.ts

Co-authored-by: cubic-dev-ai[bot] <191113872+cubic-dev-ai[bot]@users.noreply.github.com>

antman1p committed:
Fixed cubic typeguard suggestion.

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

RomanDavydchuk commented on packages/nodes-base/credentials/MicrosoftOAuth2Api.credentials.ts:
```diff
@@ -45,5 +45,24 @@ export class MicrosoftOAuth2Api implements ICredentialType {
+			],
+			default: 'https://graph.microsoft.com',
+			description: 'Select the endpoint for your Microsoft cloud environment.',
+		},
```
There are credentials that extend `microsoftOAuth2Api` but don't use the graph API:
- `AzureStorageOAuth2Api`
- `MicrosoftDynamicsOAuth2Api`
- `MicrosoftSharePointOAuth2Api`

which means that they will have this dropdown option, but it won't apply to them. This might be confusing for the user, so I suggest one of two options:
- add this option only to the credentials that use it (but this introduces code duplication)
- add `graphApiBaseUrl` property to those credentials, but with `type: 'hidden'`. This will make it hidden from the user

antman1p replied:
I've added the `graphApiBaseUrl` property with `type: 'hidden'` to all three credentials that extend `microsoftOAuth2Api` but don't use the Graph API:

- `AzureStorageOAuth2Api`
- `MicrosoftDynamicsOAuth2Api`  
- `MicrosoftSharePointOAuth2Api`

This keeps the dropdown hidden from users while maintaining compatibility with the base credential structure. The hidden field defaults to `'https://graph.microsoft.com'` to match the parent credential.

This approach avoids code duplication while ensuring users won't see a confusing option that doesn't apply to their credentials.

RomanDavydchuk reviewed (changes requested):
Hey there, I looked through the code and left a couple of comments in the code, please take a look at those. Tested the changes with MS Teams node and it seems to be working. Besides that, I noticed that for each service you have a slightly different way of getting the base URL, for example:
```typescript
	const baseUrl = (
		(typeof credentials.graphApiBaseUrl === 'string' ? credentials.graphApiBaseUrl : null) || 'https://graph.microsoft.com'
	).replace(/\/+$/, '');
```

```typescript
	const baseUrl = (
		(credentials.graphApiBaseUrl as string) || 'https://graph.microsoft.com'
	).replace(/\/+$/, '');
```

```typescript
	const baseUrl = (
		(typeof credentials.graphApiBaseUrl === 'string' && credentials.graphApiBaseUrl) || 'https://graph.microsoft.com'
	).replace(/\/+$/, '');
```

not sure why that's the case. You could see it as a nit-pick, but I'd prefer these pieces of code to be consistent and something like this makes the most sense to me:
```typescript
	const baseUrl = (
		typeof credentials.graphApiBaseUrl === 'string' && credentials.graphApiBaseUrl !== ''
			? credentials.graphApiBaseUrl
			: 'https://graph.microsoft.com'
	).replace(/\/+$/, '');
```
that way it's pretty clear that we're checking that `graphApiBaseUrl` is a non-empty string

P.s. please use a formatter when making changes to the code, I ran the CI checks and they failed because it looks like the code is not properly formatted

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Standardize graphApiBaseUrl handling and fix hardcoded URLs

antman1p commented on packages/nodes-base/nodes/Microsoft/OneDrive/GenericFunctions.ts:
I've fixed both hardcoded URLs in the OneDrive trigger:

1. **Line 48** (lastLink default value): Changed from hardcoded `'https://graph.microsoft.com/v1.0/me/drive/root/delta?token=latest'` to dynamically construct the URL using the baseUrl from credentials:
``` const credentials = await this.getCredentials('microsoftOneDriveOAuth2Api');
   const baseUrl = (
       typeof credentials.graphApiBaseUrl === 'string' && credentials.graphApiBaseUrl !== ''
           ? credentials.graphApiBaseUrl
           : 'https://graph.microsoft.com'
   ).replace(/\/+$/, '');
   
   const lastLink: string =
       (workflowData.LastLink as string) || `${baseUrl}/v1.0/me/drive/root/delta?token=latest`;
```
2. Line 75 (microsoftApiRequest call): Changed from hardcoded 'https://graph.microsoft.com/v1.0/me/drive/root/delta' to use the dynamic baseUrl:
 `${baseUrl}/v1.0/me/drive/root/delta`

Both locations now properly support different Microsoft Graph API endpoints based on the credential configuration.

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

RomanDavydchuk reviewed (changes requested):
Thanks for addressing my comments! I took a look at the code again, left just one small comment. Once this is resolved, I think we can merge this 🎉

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Add baseURL expression to requestDefaults that references
.graphApiBaseUrl, ensuring the node works with both standard and government cloud tenants.

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Fixed Test errors

antman1p committed:
Fixes the cubic-ai-dev issue on line 78 of GenericFunctions.ts.

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

RomanDavydchuk commented:
@antman1p the tests are still failing, you can try running them locally with
```bash
pnpm test
```
in `packages/nodes-base` directory

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Fixed entra test

antman1p commented:
Hi @RomanDavydchuk 

I've investigated the test failures and here's what I found:

## Summary

**All PR-related tests are passing.** The failures were due to:
1. Missing build of the `@n8n/workflow` package (now resolved)
2. One existing test that needed a mock update (now fixed)
3. 12 pre-existing failures that are completely unrelated to this PR

## Investigation Results

### ✅ All PR-Related Tests Are Passing

I verified all tests related to my changes:

```bash
cd packages/nodes-base
pnpm test Microsoft/Teams/v2/test/transport/index.test.ts \
  Microsoft/Excel/v2/test/transport/index.test.ts \
  Microsoft/Outlook/v2/test/transport/index.test.ts \
  Microsoft/OneDrive/test/GenericFunctions.test.ts \
  Microsoft/Entra/test/GenericFunctions.test.ts \
  Microsoft/ToDo/test/GenericFunctions.test.ts \
  Microsoft/GraphSecurity/test/GenericFunctions.test.ts \
  Microsoft/Entra/test/MicrosoftEntra.node.test.ts
```

**Result:** `Test Suites: 8 passed, 8 total | Tests: 103 passed, 103 total`

### Root Cause Analysis

1. **Initial 301 failures: Missing build**
   - Error: `createRunExecutionData is not a function`
   - Cause: The `@n8n/workflow` package wasn't built
   - Fix: `cd packages/workflow && pnpm build`
   - Result: Reduced from 301 to 13 failures

2. **Microsoft Entra test failure: Missing mock**
   - Error: `this.getCredentials is not a function`
   - Cause: Test didn't mock `getCredentials` (now required after adding `graphApiBaseUrl` support)
   - Fix: Added `getCredentials` mock to `MicrosoftEntra.node.test.ts`
   - Result: Test now passes ✅

3. **Remaining 12 failures: Pre-existing**
   - All are binary/file operation tests (Compression, Crypto, ReadBinaryFiles, WriteBinaryFile, etc.)
   - None use Microsoft nodes, credentials, or any code I modified
   - Verified: `pnpm test 2>&1 | grep -E "FAIL.*Microsoft"` returns **no results**

### Current Test Status

After building workflow package and fixing the one test:
- ✅ **622 test suites passing** (up from 333)
- ❌ **12 test suites failing** (down from 301) - all pre-existing
- ✅ **4,347 tests passing**
- ❌ **16 tests failing** - all pre-existing

**All 12 failing test suites:**
- `nodes/Compression/test/node/Compression.test.ts`
- `nodes/Crypto/test/Crypto.test.ts`
- `nodes/Files/ConvertToFile/test/ConvertToFile.node.test.ts`
- `nodes/Files/ExtractFromFile/test/ExtractFromFile.node.test.ts`
- `nodes/Files/ReadWriteFile/test/ReadWriteFile.test.ts`
- `nodes/Files/test/ConvertExtract.test.ts`
- `nodes/MoveBinaryData/test/MoveBinaryData.test.ts`
- `nodes/ReadBinaryFile/test/ReadBinaryFile.test.ts`
- `nodes/ReadBinaryFiles/test/ReadBinaryFiles.test.ts`
- `nodes/ReadPdf/test/ReadPDF.test.ts`
- `nodes/SpreadsheetFile/test/SpreadsheetFile.test.ts`
- `nodes/WriteBinaryFile/test/WriteBinaryFile.test.ts`

None of these are related to Microsoft Graph API or OAuth2 credential changes.

## What I Fixed

1. **Built the workflow package**: `cd packages/workflow && pnpm build`
2. **Fixed `MicrosoftEntra.node.test.ts`** by adding the `getCredentials` mock (required after adding `graphApiBaseUrl` support)

## Verification

To verify locally:
```bash
# Build workflow package first
cd packages/workflow && pnpm build

# Run all PR-related tests
cd ../nodes-base
pnpm test Microsoft/Teams/v2/test/transport/index.test.ts \
  Microsoft/Excel/v2/test/transport/index.test.ts \
  Microsoft/Outlook/v2/test/transport/index.test.ts \
  Microsoft/OneDrive/test/GenericFunctions.test.ts \
  Microsoft/Entra/test/GenericFunctions.test.ts \
  Microsoft/ToDo/test/GenericFunctions.test.ts \
  Microsoft/GraphSecurity/test/GenericFunctions.test.ts \
  Microsoft/Entra/test/MicrosoftEntra.node.test.ts

# Verify no Microsoft tests are failing
pnpm test 2>&1 | grep -E "FAIL.*Microsoft"
# Should return no results
```

## Conclusion

- ✅ All 103 PR-related tests pass
- ✅ Zero Microsoft-related test failures
- ❌ 12 pre-existing failures (unrelated to this PR)

The PR is ready for review. All functionality I added is fully tested and passing.

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

RomanDavydchuk committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

RomanDavydchuk committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'n8n-io:master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
include graphApiBaseUrl in Outlook OAuth2 test overwritten properties

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

antman1p committed:
Merge branch 'master' into fix-MS-Teams-Node-InvalidCloudInstance

RomanDavydchuk commented:
@antman1p looks like the tests are all passing now, but I'm gonna ask you to stop merging `master` into your branch now, because after each commit I need to allow workflows to run. Let's wait till all of the checks pass and then I'll merge the PR 🙏

RomanDavydchuk reviewed (approved):
LGTM

antman1p commented:
Yes, I used the client I'd and client secret. You have to make sure you
change  the .com endpoint authorization and access token urls to .us.

Then at the bottom drop down pick the correct .us graph API base url.

Also make sure the Microsoft app you created in Azure has the correct
permissions.

Let me know if it works for you.

Respectfully,

Tony


On Fri, Jan 30, 2026, 17:50 khronos-sai ***@***.***> wrote:

> *khronos-sai* left a comment (n8n-io/n8n#17297)
> <https://github.com/n8n-io/n8n/pull/17297#issuecomment-3826191025>
>
> @antman1p <https://github.com/antman1p> Hey, so we are trying this out in
> 2.6.2 in our cloud instance and notice that when you go to connect the
> account, it still redirects us to the .com addresses.
>
> Can you confirm if you used the app id and client secret route in yours?
> It doesn't seem to let us pick that if so, which is not great.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/n8n-io/n8n/pull/17297#issuecomment-3826191025>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABUSAOIQVLOSXV2COVE5BU34JPN4PAVCNFSM6AAAAACBO2UKW6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTQMRWGE4TCMBSGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

khronos-sai commented:
@antman1p Yeah, I'm able to create the application myself, we just don't have a way to connect it based on how their SaaS version works. We have nowhere to enter in a customer app ID and client secret.

antman1p commented:
Are you saying you can't enter the clientID and secret into your n8n (MS
Credential)?

On Fri, Jan 30, 2026, 18:26 khronos-sai ***@***.***> wrote:

> *khronos-sai* left a comment (n8n-io/n8n#17297)
> <https://github.com/n8n-io/n8n/pull/17297#issuecomment-3826403008>
>
> @antman1p <https://github.com/antman1p> Yeah, I'm able to create the
> application myself, we just don't have a way to connect it based on how
> their SaaS version works. We have nowhere to enter in a customer app ID and
> client secret.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/n8n-io/n8n/pull/17297#issuecomment-3826403008>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABUSAOP7BZR7IZZJUKKRRP34JPSB5AVCNFSM6AAAAACBO2UKW6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTQMRWGQYDGMBQHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

khronos-sai commented:
<img width="1376" height="180" alt="CleanShot 2026-01-30 at 15 29 31@2x" src="https://github.com/user-attachments/assets/dac4c1f6-7e5f-4579-9008-2c8ea56f9802" />


According to this, appears to be the case as well. https://docs.n8n.io/integrations/builtin/credentials/microsoftentra/#using-oauth2

antman1p commented:
OH, interesting.  I have always used the self-hosted version.  What node
are you using when you create your credential?

On Fri, Jan 30, 2026 at 6:29 PM khronos-sai ***@***.***>
wrote:

> *khronos-sai* left a comment (n8n-io/n8n#17297)
> <https://github.com/n8n-io/n8n/pull/17297#issuecomment-3826416047>
>
> @antman1p <https://github.com/antman1p> Yes, correct. In the SaaS
> version, all that is available is the "Connect my Account" button. No
> ability to define the clientID and client secret.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/n8n-io/n8n/pull/17297#issuecomment-3826416047>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABUSAOKWTP5S6ODCTV5UBST4JPSORAVCNFSM6AAAAACBO2UKW6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTQMRWGQYTMMBUG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

khronos-sai commented:
@antman1p I'm using the Entra ID node. We tested this in the Teams node as well. All seem to follow the same thing.

antman1p commented:
So when you click on "create new credential" on the Teams node, you don't
see this?

On Fri, Jan 30, 2026 at 6:39 PM khronos-sai ***@***.***>
wrote:

> *khronos-sai* left a comment (n8n-io/n8n#17297)
> <https://github.com/n8n-io/n8n/pull/17297#issuecomment-3826458426>
>
> @antman1p <https://github.com/antman1p> I'm using the Entra ID node. We
> tested this in the Teams node as well. All seem to follow the same thing.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/n8n-io/n8n/pull/17297#issuecomment-3826458426>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABUSAONJZPRQ4NG4EBCJN7T4JPTRNAVCNFSM6AAAAACBO2UKW6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTQMRWGQ2TQNBSGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

khronos-sai commented:
@antman1p I'm not seeing an image here but this is what we see. Photo attached:
<img width="2088" height="1136" alt="CleanShot 2026-01-30 at 15 43 33@2x" src="https://github.com/user-attachments/assets/ca70695f-eefe-44dc-9a9f-5cd7f605f75c" />

antman1p commented:
@khronos-sai 
<img width="946" height="596" alt="teasmcredentials" src="https://github.com/user-attachments/assets/9a4147a6-5a44-4966-8e1a-18f245139acd" />

khronos-sai commented:
@antman1p I found the issue. This is from their docs: 

<img width="1392" height="348" alt="CleanShot 2026-01-30 at 15 44 46@2x" src="https://github.com/user-attachments/assets/d77ce7d9-e78d-45f3-9224-7a273197fe13" />


And the affected lines: https://github.com/n8n-io/n8n/blob/master/packages/nodes-base/credentials/MicrosoftOAuth2Api.credentials.ts#L24-L35

These aren't exposed in the cloud version.

@RomanDavydchuk Are you able to comment on this?

antman1p commented:
Yeah, I added that to the documentation.  That's strange to me that the
cloud version would be so different.

On Fri, Jan 30, 2026 at 6:48 PM khronos-sai ***@***.***>
wrote:

> *khronos-sai* left a comment (n8n-io/n8n#17297)
> <https://github.com/n8n-io/n8n/pull/17297#issuecomment-3826507220>
>
> @antman1p <https://github.com/antman1p> I found the issue. This is from
> their docs:
> ***@***.*** (view on web)
> <https://github.com/user-attachments/assets/d77ce7d9-e78d-45f3-9224-7a273197fe13>
>
> And the affected lines:
> https://github.com/n8n-io/n8n/blob/master/packages/nodes-base/credentials/MicrosoftOAuth2Api.credentials.ts#L24-L35
>
> These aren't exposed in the cloud version.
>
> @RomanDavydchuk <https://github.com/RomanDavydchuk> Are you able to
> comment on this?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/n8n-io/n8n/pull/17297#issuecomment-3826507220>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABUSAOLYCLQ4H7SLYUZBN7T4JPUVLAVCNFSM6AAAAACBO2UKW6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTQMRWGUYDOMRSGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

khronos-sai commented:
Agreed. I'll see if we can get a comment on that here. I'm not sure if this is something (code wise) exposed to us here, so I might need to make a ticket with them.

khronos-sai commented:
@Joffcom Assuming a new PR will need to be stood up for this or is this something that the internal team will take care of?

mikenholt commented:
After following this thread for some time, I finally had the opportunity to test it today. The self-hosted version on 2.6.3 performed flawlessly in my simple test of sending a Teams message using GCC-H. I truly appreciate your hard work; it’s been incredibly valuable to us.

antman1p commented:
> After following this thread for some time, I finally had the opportunity to test it today. The self-hosted version on 2.6.3 performed flawlessly in my simple test of sending a Teams message using GCC-H. I truly appreciate your hard work; it’s been incredibly valuable to us.

@mikenholt Thank you!  I am happy it is working out for you!
